File: unix-programming/second_33/py/search-files.py
#!/usr/bin/python3

import sys
import os


ERR_INVALID_PARAM_NUMBER = 1
ERR_PARAM_NOT_DIR = 2

def eprint(*args):
    print(*args, file=sys.stderr)


def fileLines(filePath):
    return sum(1 for line in open(filePath))

def fileChars(filePath):
    return sum(len(line) for line in open(filePath))

class FileInfo(object):
    pass


if len(sys.argv) != 2:
    eprint('Invalid number of parameters!')
    sys.exit(ERR_INVALID_PARAM_NUMBER)

if sys.argv[1] == '-h':
    print('Usage: search-files.bash DIRECTORY')
    print('Searches for files with most lines and most characters')
    sys.exit(0)

if not os.path.isdir(sys.argv[1]):
    eprint("'$1' is not a directory!")
    sys.exit(ERR_PARAM_NOT_DIR)



mostLines = FileInfo()
mostLines.path = None
mostLines.amount = 0

mostChars = FileInfo()
mostChars.path = None
mostChars.amount = 0

for root, dirs, files in os.walk(sys.argv[1]):
    for fileName in files:
        filePath = os.path.join(root, fileName)
        linesAmount = fileLines(filePath)
        charsAmount = fileChars(filePath)

        if linesAmount > mostLines.amount:
            mostLines.amount = linesAmount
            mostLines.path = filePath

        if charsAmount > mostChars.amount:
            mostChars.amount = charsAmount
            mostChars.path = filePath


if mostLines.path is None and mostChars.path is None:
    print("No files found.")
else:
    print("Most lines: {0} {1}".format(mostLines.amount, mostLines.path))
    print("Most chars: {0} {1}".format(mostChars.amount, mostChars.path))

File: unix-programming/second_33/perl/search-files.pl
#!/usr/bin/perl

use strict;
use warnings;
use diagnostics;

use feature 'say';
use File::Find::Rule;


my $ERR_INVALID_PARAM_NUMBER = 1;
my $ERR_PARAM_NOT_DIR = 2;


sub get_file_stats {
    my ($file_path) = @_;
    open(INFILE, $file_path);

    my $line_count = 0;
    my $chars_count = 0;

    my $line;
    while ($line = <INFILE>) {
        $line_count++;
        $chars_count += length($line);
    }

    close(INFILE);
    return ($line_count, $chars_count);
}


my $argc = scalar @ARGV;

if ($argc >= 1 && $ARGV[0] eq "-h") {
    say 'Usage: search-files.pl DIRECTORY';
    say 'Searches for files with most lines and most characters';
    exit 0;
}

unless ($argc == 1) {
    say STDERR 'Invalid number of parameters!';
    say STDERR "Try 'search-files.pl -h' for more information.";
    exit $ERR_INVALID_PARAM_NUMBER;
}

my ($dir) = @ARGV;

unless (-d $dir) {
    say STDERR "'$dir' is not a directory!";
    exit $ERR_PARAM_NOT_DIR;
}

my @files = File::Find::Rule
    ->file()
    ->in($dir);

if (scalar(@files) eq 0) {
    say 'No files found.';
    exit 0;
}

my %most_lines_file = (
    "path" => $files[0],
    "line_count" => 0
);

my %most_chars_file = (
    "path" => $files[0],
    "chars_count" => 0
);

for (@files) {
    my ($line_count, $chars_count) = get_file_stats($_);
    if ($most_lines_file{line_count} < $line_count) {
        $most_lines_file{path} = $_;
        $most_lines_file{line_count} = $line_count;
    }
    if ($most_chars_file{chars_count} < $chars_count) {
        $most_chars_file{path} = $_;
        $most_chars_file{chars_count} = $chars_count;
    }
}

say "Most lines: $most_lines_file{line_count} $most_lines_file{path}";
say "Most chars: $most_chars_file{chars_count} $most_chars_file{path}";
